From 033cad9f9ba45f4b10a4ff30bd3abbc2180cda03 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Tue, 28 Jul 2026 16:21:07 +0800
Subject: [PATCH 435/440] AOSCOS: x86/boot: use zstd -19 if building for 32-bit

Afterglow uses native build environments (that is, 32-bit x86 chroot/
container as build environment for 32-bit x86 binaries), `zstd -22'
requests more virtual memory than that could be allowed.

Introduce `cmd_zstd_with_size' for CONFIG_X86_32 builds.

DO NOT UPSTREAM: This should really be done through build environment
detection, not by Kconfig - cross-compilation breaks the aforementioned
assumptions.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 arch/x86/boot/compressed/Makefile | 5 +++++
 scripts/Makefile.lib              | 3 +++
 2 files changed, 8 insertions(+)

diff --git a/arch/x86/boot/compressed/Makefile b/arch/x86/boot/compressed/Makefile
index 06934f9691d6a..9963ed23b8d6e 100644
--- a/arch/x86/boot/compressed/Makefile
+++ b/arch/x86/boot/compressed/Makefile
@@ -144,8 +144,13 @@ $(obj)/vmlinux.bin.lzo: $(vmlinux.bin.all-y) FORCE
 	$(call if_changed,lzo_with_size)
 $(obj)/vmlinux.bin.lz4: $(vmlinux.bin.all-y) FORCE
 	$(call if_changed,lz4_with_size)
+ifdef CONFIG_X86_32
+$(obj)/vmlinux.bin.zst: $(vmlinux.bin.all-y) FORCE
+	$(call if_changed,zstd_with_size)
+else
 $(obj)/vmlinux.bin.zst: $(vmlinux.bin.all-y) FORCE
 	$(call if_changed,zstd22_with_size)
+endif
 
 suffix-$(CONFIG_KERNEL_GZIP)	:= gz
 suffix-$(CONFIG_KERNEL_BZIP2)	:= bz2
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index 0718e39cedda2..09910dc57af3c 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -464,6 +464,9 @@ quiet_cmd_xzmisc = XZMISC  $@
 quiet_cmd_zstd = ZSTD    $@
       cmd_zstd = cat $(real-prereqs) | $(ZSTD) -19 > $@
 
+quiet_cmd_zstd = ZSTD    $@
+      cmd_zstd_with_size = { cat $(real-prereqs) | $(ZSTD) -19; $(size_append); } > $@
+
 quiet_cmd_zstd22 = ZSTD22  $@
       cmd_zstd22 = cat $(real-prereqs) | $(ZSTD) -22 --ultra > $@
 
-- 
2.55.0

